crates/unleash-edge-streaming/src/stream_broadcast.rs

```diff
@@ -0,0 +1,194 @@
+use std::pin::Pin;
+use std::sync::Arc;
+use std::time::Duration;
+
+use axum::response::{Sse, sse::Event};
+use futures_util::{Stream, StreamExt};
+use tokio::sync::RwLock;
+use tokio_stream::once;
+use tokio_stream::wrappers::BroadcastStream;
+
+use unleash_edge_delta::cache_manager::{DeltaCacheManager, DeltaCacheUpdate};
+use unleash_edge_feature_filters::{
+    FeatureFilterSet,
+    delta_filters::{DeltaFilterSet, combined_filter, filter_delta_events},
+    get_feature_filter, name_prefix_filter, project_filter_from_projects,
+};
+use unleash_edge_feature_refresh::FeatureRefresher;
+use unleash_edge_types::{
+    EdgeResult, FeatureFilters, TokenCache, errors::EdgeError, tokens::EdgeToken,
+};
+use unleash_types::client_features::ClientFeaturesDelta;
+
+fn strip_non_send(
+    result: EdgeResult<(
+        EdgeToken,
+        FeatureFilterSet,
+        unleash_types::client_features::Query,
+    )>,
+) -> EdgeResult<(EdgeToken, unleash_types::client_features::Query)> {
+    result.map(|(token, _filter_set, query)| (token, query))
+}
+
+pub async fn stream_deltas(
+    feature_refresher: Arc<FeatureRefresher>,
+    token_cache: Arc<TokenCache>,
+    edge_token: EdgeToken,
+    filter_query: FeatureFilters,
+) -> EdgeResult<Sse<impl Stream<Item = Result<Event, axum::Error>>>> {
+    let token_cache = token_cache.clone();
+    let delta_cache_manager = feature_refresher.delta_cache_manager.clone();
+
+    let (validated_token, query) = strip_non_send(get_feature_filter(
+        &edge_token,
+        &token_cache,
+        filter_query.clone(),
+    ))?;
+
+    let rx = delta_cache_manager.subscribe();
+    let streaming_query = StreamingQuery::from((&query, &validated_token));
+    let initial_features =
+        create_event_list(delta_cache_manager.clone(), 0, &streaming_query).await?;
+
+    let initial_event = Event::default()
+        .event("unleash-connected")
+        .data(serde_json::to_string(&initial_features).unwrap());
+
+    let intro_stream = once(Ok(initial_event));
+    let client_data = Arc::new(RwLock::new(ClientData {
+        revision: resolve_last_event_id(delta_cache_manager.clone(), &streaming_query),
+        streaming_query,
+    }));
+
+    let updates_stream = BroadcastStream::new(rx)
+        .take_while({
+            move |broadcast_result| {
+                let should_continue = match &broadcast_result {
+                    Ok(DeltaCacheUpdate::Deletion(_)) => false,
+                    _ => true,
+                };
+                Box::pin(async move { should_continue })
+                    as Pin<Box<dyn Future<Output = bool> + Send>>
+            }
+        })
+        .filter_map({
+            let client_data = client_data.clone();
+            let delta_cache_manager = delta_cache_manager.clone();
+            move |broadcast_result| {
+                let client_data = client_data.clone();
+                let delta_cache_manager = delta_cache_manager.clone();
```

Comment:
Do we really need to do this twice? Before and after the move?

Reply:
Yep, it looks a bit hairy but yes. There's two moves here, one into the fliter_map and one into the move |broad_cast|